add AbstractChar supertype of Char

stevengj opened the pull request:
Fixes #25302.  Wherever practical, functions that took `::Char` arguments are changed to accept any `::AbstractChar`, with fallbacks defined as necessary.

For `T<:AbstractChar`, `UInt32(c::T)` is defined to return the Unicode codepoint represented by `c` (or throw an error if `c` lies outside of Unicode), and `T(x::UInt32)` should create a `T` from the Unicode codepoint `x` (or throw an error … `T` may represent only a subset of Unicode).  This makes it possible to define generic fallbacks for comparison to `Char`, output, etcetera.    Even ancient character sets like EBCDIC have a well-defined injective mapping into Unicode, so I don't think we sacrifice any useful generality this way.

stevengj committed:
add AbstractChar supertype of Char

stevengj commented:
An exception to my comment above about injectively mapping other character sets into Unicode is the [Han unification](https://en.wikipedia.org/wiki/Han_unification), in which different character variants in Japanese encodings are mapped to the same codepoint in Unicode (which viewed them as "font" variations rather than semantically distinct characters).

If you want an `AbstractChar` type representing Shift-JIS or similar, however, I still don't see a real problem.  Conversion to `UInt32` will still produce a Unicode codepoint, since that is the only way to compare such a character to a `Char`.   You will presumably need to implement custom methods compare comparison of one JIS character to another, etcetera, but you probably would have implemented those methods anyway in order to avoid the performance cost of going through `Char`.

stevengj committed:
NEWS link

Keno commented:
One thing we had discussed was whether continuing to use `UInt32` to represent a unicode codepoint makes sense or whether, with the introduction of `AbstractChar`, we shouldn't also have `Codepoint <: AbstractChar` for that.

StefanKarpinski commented:
`CodePoint` not `Codepoint` – it's two words.

Keno commented:
SuRe ;)

nalimilan commented on base/char.jl:
```diff
@@ -24,6 +61,11 @@ function isoverlong(c::Char)
 
+# fallback: other AbstractChar types, by default, are assumed
+#           not to support malformed or overlong encodings.
+ismalformed(c::AbstractChar) = false
```
Is it really a good idea to define these fallbacks? They can be wrong for some types, and they are easy to implement. If you implement your own `AbstractChar` type, it's good to think about this question precisely.

stevengj replied:
All of the proposed `AbstractChar` types I've ever seen, other than `Char`, don't support malformed characters.   So it makes sense to me to make this the default.  You can always override it.

Another reason to make this the default is that `ismalformed(c)` returns `true`, you need to define additional methods to decode the character.

nalimilan commented on base/char.jl:
```diff
@@ -69,50 +111,57 @@ function Char(b::Union{Int8,UInt8})
+isless(x::AbstractChar, y::AbstractChar) = isless(Char(x), Char(y))
+==(x::AbstractChar, y::AbstractChar) = Char(x) == Char(y)
+hash(x::AbstractChar, h::UInt) =
+    hash_uint64(((UInt32(x) + UInt64(0xd060fad0)) << 32) ⊻ UInt64(h))
```
Why not use `reinterpret`, so that invalid codepoints can be hashed (just like for `Char`)?

Keno replied:
`AbstractChar` need not be reinterpretable to UInt32.

nalimilan replied:
But you could convert to `Char` first...

Keno replied:
Sure, that's a good point I guess, since equality is defined in terms of comparison, to `Char`, hashing should match.

nalimilan commented on base/strings/basic.jl:
```diff
@@ -13,9 +13,9 @@ about strings:
-  * Only the index of the first code unit of a `Char` is a valid index
-  * The encoding of a `Char` is independent of what precedes or follows it
+  * Each `AbstractChar` in a string is encoded by one or more code units
+  * Only the index of the first code unit of a `AbstractChar` is a valid index
```
"an". Same below.

nalimilan commented on base/strings/basic.jl:
```diff
@@ -508,17 +508,18 @@ false
+            "map(f, s::AbstractString) requires f to return AbstractChar; " *
             "try map(f, collect(s)) or a comprehension instead"))
-        write(out, c′::Char)
+        write(out, c′::AbstractChar)
```
Do we require all `AbstractChar` implementations to implement `write` so that it uses UTF-8? This code relies on this assumption.

stevengj replied:
There is a fallback `write` method that works via conversion to `Char`.

nalimilan replied:
OK. But I guess it should be mentioned in the docstring for `AbstractChar` too? It could sound natural to somebody implementing e.g. a `Latin1Char` to write it in ISO-8859-1, which wouldn't be correct.

stevengj replied:
Will do.

stevengj replied:
Alternatively, we could define `print` to always output UTF-8, and `write` to output a raw encoded value.

stevengj replied:
But I tend to think that the I/O encoding should be a property of the stream, not the string or character type.

StefanKarpinski replied:
100% agree: letting the value that's being printed determine the encoding would be a mess. We do need external support for encoded streams that handle this appropriately, but `print(io, '∀')` where `io` is a UTF-16 encoded stream it should definitely not write UTF-8 data to `io` because `'∀'` is a UTF-8 encode character type. If you have a `CharU16` value and a UTF-16 encoded stream it should know that it can output code units directly, of course, but that's just a matter of providing the right `print` specializations.

nalimilan commented on base/strings/util.jl:
```diff
@@ -410,7 +410,7 @@ If `count` is provided, replace at most `count` occurrences.
 If `r` is a function, each occurrence is replaced with `r(s)`
 where `s` is the matched substring (when `pat`is a `Regex` or `AbstractString`) or
-character (when `pat` is a `Char` or a collection of `Char`).
+character (when `pat` is a `AbstractChar` or a collection of `AbstractChar`).
```
"an"

nalimilan commented on doc/src/manual/strings.md:
```diff
@@ -29,8 +29,7 @@ There are a few noteworthy high-level features about Julia's strings:
   * Like C and Java, but unlike most dynamic languages, Julia has a first-class type representing
-    a single character, called `Char`. This is just a special kind of 32-bit primitive type whose numeric
-    value represents a Unicode code point.
+    a single character, called `AbstractChar`. This is just a special kind of 32-bit primitive type whose numeric value represents a Unicode code point.
```
Not necessarily 32-bit now. Maybe mention `Char` or say "by default"?

nalimilan commented on base/char.jl:
```diff
@@ -1,12 +1,49 @@
+"""
+Char
+
+struct InvalidCharError{T<:AbstractChar} <: Exception
```
Maybe `InexactError` could be used here? Just an idea.

stevengj replied:
Possibly, but I'm not sure if that change belongs in this PR.

nalimilan replied:
Sorry, I hadn't realized this type already existed in master.

stevengj commented:
I'm not sure what `CodePoint` accomplishes?

Keno commented:
The idea of codepoint is to make clear what the conversion is about. `convert(UInt32, c)` works fine technically of course, but the question is do you really want that? E.g. do you want to be able to `push!(UInt32[1], 'a')`? I'd argue that that operation doesn't really make much sense. One is an integer the other is a codepoint. They don't really have the same operations (e.g. you can do arithmetic on integers, but not on codepoints). It just seems like much larger of a semantic bridge than we usually ascribe to `convert`. A related, but separate argument is that packages will want a UTF32-style string anyway, for which `CodePoint` is the natural character type, so it makes sense to pre-define it for operations that operate on codepoints.

stevengj commented:
@Keno, we could define the constructors `Char(int)` and `UInt32(Char)` and not `convert` to avoid accidental conversions.  That seems like a sensible thing to me, is consistent with e.g. #16024, and would address your objection without introducing a new type.

Rather than defining a single new type, it makes more sense to me to have `codepoint(c::AbstractChar)::Integer` return the codepoint of `c` in the most appropriate integer type for `c` (e.g. `UInt8` for ASCII), and have `codepoint(::Type{<:AbstractChar})` return the corresponding integer type.

stevengj committed:
fixes from feedback

StefanKarpinski commented on base/char.jl:
```diff
@@ -69,50 +112,56 @@ function Char(b::Union{Int8,UInt8})
+convert(::Type{T}, x::AbstractChar) where {T<:Number} = T(x)
 
-rem(x::Char, ::Type{T}) where {T<:Number} = rem(UInt32(x), T)
+rem(x::AbstractChar, ::Type{T}) where {T<:Number} = rem(UInt32(x), T)
```
It seems a bit off that this supports any kind of `Number` rather than just `Integer`. Pre-existing issue, I know.

StefanKarpinski commented on base/char.jl:
```diff
@@ -69,50 +112,56 @@ function Char(b::Union{Int8,UInt8})
-+(x::Integer, y::Char) = y + x
+# fallbacks:
+isless(x::AbstractChar, y::AbstractChar) = isless(Char(x), Char(y))
+==(x::AbstractChar, y::AbstractChar) = Char(x) == Char(y)
```
It seems better for the fallback comparisons to be done in `UInt32`.

stevengj replied:
I originally had it that way, but this way is more general.  The issue is that all Unicode codepoints can be represented by `Char`, but not vice versa.

Converting to `UInt32` would mean that `ASCIIChar('a') == typemax(Char)` would throw an `InvalidCharError` rather than returning `false`.

StefanKarpinski commented on base/char.jl:
```diff
@@ -69,50 +112,56 @@ function Char(b::Union{Int8,UInt8})
 const hex_chars = UInt8['0':'9';'a':'z']
 
-function show(io::IO, c::Char)
+function show(io::IO, c::AbstractChar)
```
I don't think this is generic because of the call to `reinterpret(UInt32, c)` – we have no idea what the representation of an abstract character type is – it may not even be 32-bits or a bits type at all. We could extract `show_invalid(c::AbstractChar)` as a method that characters have to implement.

nalimilan replied:
As noted in another thread, we could use `reinterpret(UInt32, Char(c))`.

stevengj replied:
Yes, the `reinterpret` call is only for invalid chars.   I agree that refactoring this seems like a good choice.

StefanKarpinski commented on base/strings/basic.jl:
```diff
@@ -140,7 +140,7 @@ See also: [`getindex`](@ref), [`start`](@ref), [`done`](@ref),
 start(s::AbstractString) = 1
 done(s::AbstractString, i::Integer) = i > ncodeunits(s)
-eltype(::Type{<:AbstractString}) = Char
+eltype(::Type{<:AbstractString}) = Char # some string types may use another AbstractChar
```
Is there some way we can put an `::eltype(s)` type assert somewhere to catch this if it's wrong?

stevengj replied:
Not sure…where would it go?

StefanKarpinski commented on base/strings/unicode.jl:
```diff
@@ -634,8 +634,8 @@ graphemes(s::AbstractString) = GraphemeIterator{typeof(s)}(s)
-function length(g::GraphemeIterator)
-    c0 = typemax(Char)
+function length(g::GraphemeIterator{S}) where {S}
+    c0 = eltype(S)(0x00000000)
```
Is `'\0'` a safe character for this? It's not a  malformed character – does the Unicode spec ensure that it can't ever combine with anything?

stevengj replied:
Yes.

stevengj replied:
(I used `\0` here since hopefully every conceivable `AbstractChar` type can encode this.)

StefanKarpinski commented on doc/src/manual/strings.md:
```diff
@@ -28,9 +28,9 @@ There are a few noteworthy high-level features about Julia's strings:
-    value represents a Unicode code point.
+  * Like C and Java, but unlike most dynamic languages, Julia has a first-class type for representing
+    a single character, called `AbstractChar`. The built-in `Char` subtype of `AbstractChar`
+    is a 32-bit primitive type that can represent any Unicode character.
```
Perhaps mention that `Char` represents Unicode characters as zero-padded UTF-8 bytes rather than as code point values? This seems like as good a place as any to mention that.

StefanKarpinski commented on doc/src/manual/strings.md:
```diff
@@ -41,9 +41,10 @@ There are a few noteworthy high-level features about Julia's strings:
 ## [Characters](@id man-characters)
 
-A `Char` value represents a single character: it is just a 32-bit primitive type with a special literal
+An `AbstractChar` value represents a single character: it is just a 32-bit primitive type with a special literal
```
No longer accurate since an `AbstractChar` could have any representation at all. The simplest fix is just to leave this as `Char`.

StefanKarpinski reviewed (approved):
Very nice. I approve modulo comments.

stevengj commented:
`write(io, 0x00)` and `write(io, 0x0000)` output the same (`isequal`) value in different "encodings", whereas `print` in both cases outputs `0`.   So maybe it should be the same for `AbstractChar`.  (This is also how `UTF16String` was read and written by `read`/`write` in LegacyStrings, and it seemed pretty convenient).

`print(io, x)` is defined to output text, and in that case I agree that the encoding should be determined by `io` (defaulting to UTF-8) and not by `x`.

(So, there would be a fallback `print(io, x::AbstractChar) = print(io, Char(x))`, but no fallback for `write` or `read`: implementations would be responsible for providing the latter.)

StefanKarpinski commented:
re: `print`/`write` behavior, that seems reasonable to me 👍. @JeffBezanson may have thoughts.

stevengj committed:
use print when io-specific text encoding (usually UTF-8) is desired, and write when encoding should be determined by the argument type

stevengj committed:
restore optimized print for strings, and analogous optimizations for SubString{String}

stevengj commented:
I updated the PR to make the `print`/`write` distinction discussed above.

I also did a pass over the source code and I found lots of `write` calls that should be `print` calls if we want the latter to (potentially) have an `io`-specific encoding.  I fixed most of these, since there should be no performance difference.   As an exception, there are various cases (e.g. in printing numbers) where we output raw ASCII bytes to the stream with `write` or `unsafe_write`.  I left these as-is so as not to affect performance, but future revisions should think about how to make this more generic.   Improving this in the future shouldn't require breaking changes, though, so it can wait until 1.1 or later.

stevengj committed:
updates in response to documentation comments

stevengj commented:
@Keno, I tried getting rid of the `convert` functions and ran into trouble right away: `base/chars.jl` uses `const hex_chars = UInt8['0':'9';'a':'z']`, which implicitly relies on `convert(UInt8, ::Char)`.   Alternatives like `UInt8.(['0':'9';'a':'z'])` seem much more annoying.   So, I'd prefer to leave that change to another PR, if any.

Keno commented:
I'm actually much more comfortable with an implicit conversion to `UInt8`, as long as that conversion is only defined where `convert(UInt8, c::Char) == reinterpret(UInt32, c::Char) % UInt8`.

Keno commented:
But fine to do in a different PR.

stevengj committed:
add codepoint function and more conversions, and move some stuff from boot.jl to char.jl

stevengj commented:
Added the `codepoint(c::Char)` function as discussed above.

StefanKarpinski commented on base/char.jl:
```diff
@@ -16,15 +18,14 @@ The [`isvalid`](@ref) function can be used to check which codepoints are
-their own implementations of `write` and `read`.
+via `codepoint(char)` will not reveal this encoding because it always returns the
+Unicode value of the character. `print(io, c)` of any `c::AbstractChar`
+produces UTF-8 output by default (via conversion to `Char` if necessary).
```
Perhaps clarify that it's not necessary that `print(io, c)` produces UTF-8 but rather that the output encoding is determined by `io` – and the built-in `IO` types are all UTF-8.

stevengj commented:
Note that to really support non-UTF8 `IO` streams, we'll eventually need to have some kind of `trait(io)` that gives us information about the encoding.

Not only do we want to know whether it is UTF-8 (so that e.g. the fast path for `print(io, ::String)` can be used, but we also want to know whether it is any superset of ASCII (so that e.g. the fast path for writing ASCII numeric digits can be used).

That will involve a separate design process, and should be non-breaking as far as I can tell, so it can happen outside of this PR.

stevengj committed:
more tests, fixes

stevengj commented:
AppVeyor CI failure seems to be an unrelated timeout.
